GB ASM Tutorial Cheatsheet (#68)
gbdev/gb-asm-tutorial: src/cheatsheet.md

# RGBDS Cheatsheet

> The purpose of this page is to provide concise explanations and code-snippets for common tasks.

Reviewer: ```suggestion
The purpose of this page is to provide concise explanations and code snippets for common tasks.
```

For extra depth, clarity, and understanding; it's recommended you read through the [Hello World](part1/setup.md), [Part II - Our first game](part2/getting-started.md), and [Part III - Our second game](part3/getting-started.md) tutorials.

Assembly syntax & CPU Instructions will not be explained, for more information see the [RGBDS Language Reference](https://rgbds.gbdev.io/docs/v0.6.1/rgbasm.5)

> Is there something common you think is missing? Reach out on the @gbdev discord server for any suggestions.

Reviewer: ```suggestion
Is there something you think is missing? Check the [github repository](https://github.com/gbdev/gb-asm-tutorial) to open an Issue or contribute to this page.
```

Author: Maybe both? Mentioning the gbdev discord server AND the ability to open issues themselves?

ldh [rLCDC], a
```

> **NOTE:** Sprites by default are in 8x8 mode. 
### How to enable tall (8x16) sprites

Once sprites are enabled, you can enable tall sprites using the third least signficiant bit of the `rLCDC` register: `LCDCF_OBJ16` 

> **NOTE:** You can not have some 8x8 sprites and some 8x16 sprites. All sprites must be of the same size.

```rgbasm,linenos
; Turn tall sprites on
ld a, [rLCDC]
or a, LCDCF_OBJ16
ldh [rLCDC], a
```

## Backgrounds

### How to put background/window tile data into VRAM

The reigon in VRAM dedicated for the background/window tilemaps is from $9000 to $97F0. Harware.inc defines a `_VRAM9000` constant you can use for that. To copy background/window tile data into VRAM, you can use a loop to copy the bytes.

myBackground: INCBIN "src/path/to/my/my-background.2bpp"
myBackgroundEnd:

CopyBackgroundWindowTileDataIntoVram:
    ; Copy the tile data
    ld de, myBackground
    ld hl, _VRAM
    ld bc, myBackgroundEnd - myBackground
CopyBackgroundWindowTileDataIntoVram_Loop:
    ld a, [de]
    ld [hli], a
    inc de
    dec bc
    ld a, b
    or a, c
    jp nz, CopyBackgroundWindowTileDataIntoVram_Loop

### How to Draw on the Background/Window

> The Game Boy has 2 32x32 tilemaps, once at `$9800` and another at `$9C00`. Either can be used for the background or window. By default, they both use the tilemap at `$9800`. 

Reviewer: ```suggestion
The Game Boy has 2 32x32 tilemaps, one at `$9800` and another at `$9C00`. Either can be used for the background or window. By default, they both use the tilemap at `$9800`. 
```

> Drawing on the background or window is as simple as copying bytes starting at one of those addresses:

Reviewer: this feels like something should follow, but then there are two notes? Is something missing?

Author: it's referencing the code snippet after the notes. maybe moving the notes after the snippet?

Reviewer: yep, agree

Author: Moved

**Usage Explanation (From the Github):**

The library is relatively simple to get set up. First, put the following in your initialization code:

```rgbasm, linenos
	; Initilize Sprite Object Library.
	call InitSprObjLib

	; Reset hardware OAM
	xor a, a
	ld b, 160
	ld hl, _OAMRAM
.resetOAM
	ld [hli], a
	dec b
	jr nz, .resetOAM
```
Then put a call to `ResetShadowOAM` at the beginning of your main loop.

Finally, run the following code during VBlank:

```rgbasm, linenos
ld a, HIGH(wShadowOAM)
call hOAMDMA
```

### How to manipulate Shadow OAM OAM sprites

Once you've setup @eievue5's Sprite Object Library, you can manipulate shadow OAM sprites the exact same way you would manipulate normal hardware OAM sprites. Except, this time you would use the librarys `wShadowOAM` constant instead of the `_OAMRAM` register.

**Moving (the first) OAM sprite, one pixel downwards:**

```rgbasm, linenos
ld a,LOW(wShadowOAM)
ld l, a
ld a, HIGH(wShadowOAM)
ld h, a

ld a, [hl]
inc a
ld [wShadowOAM], a
```

> ## Miccelaneous

Author: Done

Here are some other common topics that developers will commonly run into.
### How to Save Data

> TODO - i'm so ignorant on this topic, i'll place the little i've learned anyways.

Author: i filled out this section to the best of my ability, if nobody has the time to review, i can just remove for later.

### How to wait for the vertical blank phase

To check for the vertical blank phase, use the `rLY` register. Compare that register's value against the height of the Game Boy screen in pixels: 144.

```rgbasm,linenos
WaitUntilVerticalBlankStart:
    ld a, [rLY]
    cp 144
    jp nc, WaitUntilVerticalBlankStart
```

To wait until the vertical blank phase is finished, you can use a similar code snippet:

```rgbasm,linenos
WaitUntilVerticalBlankEnd:
    ld a, [rLY]
    cp 144
    jp c, WaitUntilVerticalBlankEnd
```

The only difference being the condition for the jump command.

### How to turn on/off the LCD Display

You can turn the LCD on and off by altering the most significant bit controls the state of the `rLCDC` register. Hardware.inc also has constants for altering that bit: `LCDCF_ON` and `LCDCF_OFF`.

**To turn the LCD on:**
	
```rgbasm,linenos
ld a, LCDCF_ON
ldh [rLCDC], a
```
**To turn the LCD off:**

> **NOTE:** Do not turn the LCD off outside of the Vertical Blank Phase. See "[How to wait for vertical blank phase](#how-to-wait-for-the-vertical-blank-phase)".
	
```rgbasm,linenos
; Turn the LCD off
ld a, LCDCF_OFF
ldh [rLCDC], a
```

### How to turn on/off the background

> To turn the background layer on and off, alter the least significant bit of the `rLCDC` register. You can use the `LCDCF_BGON` and `LCDCF_BGOFF` constants for this.

Author: > Note also that this has different behaviour on DMG and CGB, and it doesn't allow access to VRAM even when off; so I don't think this section is much use at all?

Should i remove this section?

> To turn the background layer on and off, alter the least significant bit of the `rLCDC` register. You can use the `LCDCF_BGON` and `LCDCF_BGOFF` constants for this.

Reviewer 2: Yeah, I think that's better. I'd assume people interested will read Pan Docs, and figure out how to modify the other, similar snippets. It shouldn't hurt to focus on the most useful ones rather than try to be exhaustive, for a cheat sheet.

> To turn the background layer on and off, alter the least significant bit of the `rLCDC` register. You can use the `LCDCF_BGON` and `LCDCF_BGOFF` constants for this.

Author: > Note also that this has different behaviour on DMG and CGB, and it doesn't allow access to VRAM even when off; so I don't think this section is much use at all?

It's an easy remove , but What's the behavior difference? I guess part of me feels like turning on/ off the window at least is common.

    - [How to Draw on the Background/Window](#how-to-draw-on-the-backgroundwindow)
    - [How to move the background](#how-to-move-the-background)
    - [How to move the window](#how-to-move-the-window)
  - [Joypad Input](#joypad-input)
    - [Checking if a button is down](#checking-if-a-button-is-down)
    - [Checking if a button was JUST pressed](#checking-if-a-button-was-just-pressed)
    - [How to wait for a button press](#how-to-wait-for-a-button-press)
  - [HUD](#hud)
    - [How to Draw Text](#how-to-draw-text)
    - [How To draw a bottom HUD](#how-to-draw-a-bottom-hud)
  - [Sprites](#sprites)
    - [How to put sprite tile data in VRAM](#how-to-put-sprite-tile-data-in-vram)
    - [How to manipulate hardware OAM sprites](#how-to-manipulate-hardware-oam-sprites)
    - [How to implement a Shadow OAM using @eievue5's Sprite Object Library](#how-to-implement-a-shadow-oam-using-eievue5s-sprite-object-library)
    - [How to manipulate Shadow OAM OAM sprites](#how-to-manipulate-shadow-oam-oam-sprites)
  - [Miccelaneous](#miccelaneous)
    - [How to Save Data](#how-to-save-data)
    - [How to generate random numbers](#how-to-generate-random-numbers)

## Display

> The `$FF40` register controls all of the following:

Author: I've got some littered in there. I definitely could use more.

> The `$FF40` register controls all of the following:

Author: adding some more pan docs links.

- The LCD display
- the background
- the window
- sprites
- tall sprites

In hardware.inc, you can find a constant for that register: `rLCDC`. For more information on LCD control, refer to the [Pan Docs](https://gbdev.io/pandocs/LCDC.html)

### How to wait for the vertical blank phase

> To check for the vertical blank phase, use the `rLY` register. Compare that register's value against the height of the Game Boy screen in pixels: 144.

Author: Let's find the most common way, and then maybe leave links and notes about where to find info for the other ways and documentation about the technical differences?

> To check for the vertical blank phase, use the `rLY` register. Compare that register's value against the height of the Game Boy screen in pixels: 144.

Reviewer 2: I agree with the sentiment. But for this one specifically, even that won't work..!

The two common ways are LY and VBlank handler, and the latter is inherently dependent on said handler. The former is used mostly when the handler hasn't been set up yet (i.e. when turning off the LCD during boot-up), and the latter everywhere else.

> To check for the vertical blank phase, use the `rLY` register. Compare that register's value against the height of the Game Boy screen in pixels: 144.

Author: > I agree with the sentiment. But for this one specifically, even that won't work..!
> 
> The two common ways are LY and VBlank handler, and the latter is inherently dependent on said handler. The former is used mostly when the handler hasn't been set up yet (i.e. when turning off the LCD during boot-up), and the latter everywhere else.

Yeahhh... i don't understand that :D

> To check for the vertical blank phase, use the `rLY` register. Compare that register's value against the height of the Game Boy screen in pixels: 144.

Reviewer: Can we introduce the three ways and just explain/describe one for the time being?

> To check for the vertical blank phase, use the `rLY` register. Compare that register's value against the height of the Game Boy screen in pixels: 144.

Reviewer 3: I don't think the vblank handler fits in the same entry. This is how to do nothing *except* wait for vblank.

> To check for the vertical blank phase, use the `rLY` register. Compare that register's value against the height of the Game Boy screen in pixels: 144.

Author: What's a good gameplan here? I don't know "the three ways"

> I don't think the vblank handler fits in the same entry. This is how to do nothing _except_ wait for vblank.

That's the purpose of the snippet?

> To check for the vertical blank phase, use the `rLY` register. Compare that register's value against the height of the Game Boy screen in pixels: 144.

Reviewer 3: Sorry, I was perhaps a bit too brief in my previous comment. I'm responding to the debate/confusion surrounding which technique should be highlighted here.

Briefly:
- the `LY` way (as it is in the text currently) is the best candidate for "how to wait for VBlank".
- the VBlank handler is an important topic that should be covered, but is a separate thing that should be treated as such (e.g. a separate entry / under a different heading)
- It's easier to understand `LY` than using `STAT` -- loosely: scanlines are "real" but PPU is a black box.

Not briefly:
- I think it's reasonable to say that *the way* (recommended way, in general, etc.) to **wait for VBlank** is to enter a busy loop and read `LY` until scanline 144. This is "doing nothing *except* waiting".
  - This is essentially what the text is already saying. The snippet is fine.
  - Ideally the text will not give too strong an impression that that's the only way to do anything re. VBlank.
- My point about the VBlank interrupt handler is that it doesn't need to be considered as a competing technique to `LY144` as the usual point of using it is significantly different to *waiting* for VBlank. They serve different purposes.
  - The VBlank interrupt is for efficient game loop synchronisation in concert with `halt` (in general, etc.)
  - I would recommend **not** using a busy loop to synchronise a game loop.
- verbose: The implication of doing nothing except waiting is that the program has nothing else to do, or nothing better to do at that time (or no other reliable means to "V-sync"). Using `halt` to wait for the VBlank interrupt and save power would be an example of something better to do, if all the program is doing is delaying until the next frame (synchronising the game loop).

> To check for the vertical blank phase, use the `rLY` register. Compare that register's value against the height of the Game Boy screen in pixels: 144.

Author: I guess I sort of get what youre saying... Maybe that should be an addition later down the line?  Unless someone else wants to submit the code snippet...

> To check for the vertical blank phase, use the `rLY` register. Compare that register's value against the height of the Game Boy screen in pixels: 144.

Author: > I think it shouldn't be a code snippet, then; but that this one's description should at least hint at the use of `halt` instead of this, whenever available. Perhaps with a link to the relevant section of the tutorial that explains that (if there isn't one yet, that's fine, and we'll open an issue to remind us later).

Sure, I can do that. I'm not familiar with `halt`, can you give me some documentation links or a snippet to add?

```rgbasm,linenos
WaitUntilVerticalBlankStart:
    ld a, [rLY]
    cp 144

>     jp nc, WaitUntilVerticalBlankStart

Author: fixing

```

To wait until the vertical blank phase is finished, you can use a similar code snippet:

```rgbasm,linenos
WaitUntilVerticalBlankEnd:
    ld a, [rLY]
    cp 144
    jp c, WaitUntilVerticalBlankEnd
```

> The only difference being the condition for the jump command.

Author: will remove the snippet with an explanation

### How to turn on/off the LCD Display

You can turn the LCD on and off by altering the most significant bit controls the state of the `rLCDC` register. Hardware.inc also has constants for altering that bit: `LCDCF_ON` and `LCDCF_OFF`.

**To turn the LCD on:**
	
```rgbasm,linenos

> ; Turn the LCD on

Reviewer 2: I think it's best to avoid comments redundant with the main text; though maybe sometimes that'd mean removing the main text instead.

```suggestion
```

# RGBDS Cheatsheet

The purpose of this page is to provide concise explanations and code snippets for common tasks.
For extra depth, clarity, and understanding, it's recommended you read through the [Hello World](part1/setup.md), [Part II - Our first game](part2/getting-started.md), and [Part III - Our second game](part3/getting-started.md) tutorials.

> Assembly syntax & CPU Instructions will not be explained, for more information see the [RGBDS Language Reference](https://rgbds.gbdev.io/docs/v0.6.1/rgbasm.5)

Reviewer: @ISSOtm can you help here?

> Assembly syntax & CPU Instructions will not be explained, for more information see the [RGBDS Language Reference](https://rgbds.gbdev.io/docs/v0.6.1/rgbasm.5)

Reviewer 2: ![image](https://github.com/gbdev/gb-asm-tutorial/assets/15271137/1f36bb98-634d-49bf-95fa-979363848af8)

gdi it didn't send. \*mutters angrily about GitHub's interface\*

Is there something common you think is missing? Check the [github repository](https://github.com/gbdev/gb-asm-tutorial) to open an Issue or contribute to this page. Alternatively, you can reach out on one of the @gbdev [community channels](https://gbdev.io/chat.html).

## Table of Contents

-   [Display](#display)
    -   [Wait for the vertical blank phase](#wait-for-the-vertical-blank-phase)
    -   [Turn on/off the LCD Display](#turn-onoff-the-lcd-display)
    -   [Turn on/off the background](#turn-onoff-the-background)
    -   [Turn on/off the window](#turn-onoff-the-window)
    -   [Making the window and background use different tilemaps](#making-the-window-and-background-use-different-tilemaps)
    -   [Turn on/off sprites](#turn-onoff-sprites)
    -   [Enable tall (8x16) sprites](#enable-tall-8x16-sprites)
-   [Backgrounds](#backgrounds)
    -   [Put background/window tile data into VRAM](#put-backgroundwindow-tile-data-into-vram)
    -   [Draw on the Background/Window](#draw-on-the-backgroundwindow)
    -   [Move the background](#move-the-background)
    -   [Move the window](#move-the-window)
-   [Joypad Input](#joypad-input)

>     -   [Checking if a button is down](#checking-if-a-button-is-down)

Author: done. i changed 3 titles. the "checking if a button is down/was pressed". and the "making the window and background use different tilemaps"

    -   [Manipulate Shadow OAM OAM sprites](#manipulate-shadow-oam-oam-sprites)
-   [Micelaneous](#micelaneous)
    -   [Save Data](#save-data)
    -   [Generate random numbers](#generate-random-numbers)

## Display

The `$FF40` register controls all of the following:

-   The LCD display
-   the background
-   the window
-   sprites
-   tall sprites

In hardware.inc, you can find a constant for that register: `rLCDC`. For more information on LCD control, refer to the [Pan Docs](https://gbdev.io/pandocs/LCDC.html)

### Wait for the vertical blank phase

To check for the vertical blank phase, use the `rLY` register. Compare that register's value against the height of the Game Boy screen in pixels: 144.

```rgbasm,linenos
WaitUntilVerticalBlankStart:
    ld a, [rLY]
    cp 144
    jp c, WaitUntilVerticalBlankStart
```

::: tip

To wait until the vertical blank phase is finished, you would use a code-snippet like the one above. Except you would continue looping until `rLY` is **less than** 144 (when `cp 144` has no carry-over)

:::

### Turn on/off the LCD Display

> You can turn the LCD on and off by altering the most significant bit controls the state of the `rLCDC` register. Hardware.inc also has constants for this: `LCDCF_ON` and `LCDCF_OFF`.

Reviewer 4: ```suggestion
You can turn the LCD on and off by altering the most significant bit of the `rLCDC` register. hardware.inc has constants for this: `LCDCF_ON` and `LCDCF_OFF`.
```

**To turn the LCD on:**

```rgbasm,linenos
ld a, LCDCF_ON
ldh [rLCDC], a
```

**To turn the LCD off:**

::: warning

Do not turn the LCD off outside of the Vertical Blank Phase. See "[How to wait for vertical blank phase](#how-to-wait-for-the-vertical-blank-phase)".

:::

```rgbasm,linenos
; Turn the LCD off
ld a, LCDCF_OFF
ldh [rLCDC], a
```

### Turn on/off the background

To turn the background layer on and off, alter the least significant bit of the `rLCDC` register. You can use the `LCDCF_BGON` and `LCDCF_BGOFF` constants for this.

> **To turn the Background on:**

Reviewer 4: Consistent capitalization.
```suggestion
**To turn the background on:**
```

```rgbasm,linenos
; Turn the background on

> ld a, [rLCDC]

Reviewer 4: Please use `ldh` whenever possible. Make sure you go over the document for these; this is only one instance.
```suggestion
ldh a, [rLCDC]
```

or a, LCDCF_BGON
ldh [rLCDC], a
```

> **To turn the Background off:**

Reviewer 4: Consistent capitalization.
```suggestion
**To turn the background off:**
```

```rgbasm,linenos
; Turn the background off
ld a, [rLCDC]
and a, LCDCF_BGOFF
ldh [rLCDC], a
```

### Turn on/off the window

To turn the window layer on and off, alter the least significant bit of the `rLCDC` register. You can use the `LCDCF_WINON` and `LCDCF_WINOFF` constants for this.

> **To turn the Window on:**

Reviewer 4: Consistent capitalization.
```suggestion
**To turn the window on:**
```

```rgbasm,linenos
; Turn the window on
ld a, [rLCDC]
or a, LCDCF_WINON
ldh [rLCDC], a
```

> **To turn the Window off:**

Reviewer 4: Consistent capitalization.
```suggestion
**To turn the window off:**
```

```rgbasm,linenos
; Turn the window off
ld a, [rLCDC]
and a, LCDCF_WINOFF
ldh [rLCDC], a
```

### Making the window and background use different tilemaps

> By default, the window and background layeer will use the same tilemap. That is, any tiles you draw on the background/window, will be mirrored on the window/background.

Reviewer 5: spelling; reword without an implicit "respectively"
```suggestion
By default, the window and background layer will use the same tilemap. That is, any tiles you draw on the background will be mirrored on the window and vice versa.
```

> For the window & background, there are 2 memory spaces they can utilize: `$9800` and `$9C00`. For more information, refer to the [Pan Docs](https://gbdev.io/pandocs/Tile_Maps.html)

Reviewer 5: nit: write out "and"; "use" is simpler than "utilize"
```suggestion
For the window and background, there are 2 memory spaces they can use: `$9800` and `$9C00`. For more information, refer to the [Pan Docs](https://gbdev.io/pandocs/Tile_Maps.html)
```

> For the window and background, there are 2 memory spaces they can use: `$9800` and `$9C00`. For more information, refer to the [Pan Docs](https://gbdev.io/pandocs/Tile_Maps.html)
> 
> Which space the background uses is controled by the fourth **least** significant bit of the `rLCDC` register. Which page the window uses is controlled by the 2 **most** significant bit.
> 
> You can use one of the 4 constants to specify which layer uses which space:

Reviewer 4: Memory region, not space.

Also, please refer to bits by number, not significance. it's more consistent with other documentation, diagrams, and libraries, like hardware.inc.
```suggestion
For the window and background, there are 2 memory regions they can use: `$9800` and `$9C00`. For more information, refer to the [Pan Docs](https://gbdev.io/pandocs/Tile_Maps.html)

Which region the background uses is controlled by the 4th bit of the `rLCDC` register. Which region the window uses is controlled by the 7th bit.

You can use one of the 4 constants to specify which layer uses which region:
```

-   LCDCF_WIN9800
-   LCDCF_WIN9C00
-   LCDCF_BG9800
-   LCDCF_BG9C00

::: tip Note

You still need to make sure the window and background are turned on when using these constants.

:::

> 

Author: > This is a lot of detail information. This entry has no snippet.

is it critical to have a snippet? it's just explaining the purpose of the section

> 

Reviewer 3: I'm not saying that it has to have a snippet necessarily, it just seemed to be less of a concise cheatsheet tip than the other sections.

I guess I don't have any specific, actionable advice and the text isn't wrong, so you can probably ignore this.

### Turn on/off sprites

> Sprites can be toggled on and off using the second least significant bit of the `rLCDC` register. You can use the `LCDCF_OBJON` and `LCDCF_OBJOFF` constants for this.

Reviewer 5: Some tutorials refer to them as "objects", others as "sprites". Use both to help readers understand that the terms are synonyms.
```suggestion
Sprites (or objects) can be toggled on and off using the second least significant bit of the `rLCDC` register. You can use the `LCDCF_OBJON` and `LCDCF_OBJOFF` constants for this.
```

> Sprites (or objects) can be toggled on and off using the second least significant bit of the `rLCDC` register. You can use the `LCDCF_OBJON` and `LCDCF_OBJOFF` constants for this.

Reviewer 4: ```suggestion
Sprites (or objects) can be toggled on and off using the 2nd bit of the `rLCDC` register. You can use the `LCDCF_OBJON` and `LCDCF_OBJOFF` constants for this.
```

> **To turn the Sprites On:**

Reviewer 4: ```suggestion
**To turn sprite objects on:**
```

```rgbasm,linenos
; Turn the sprites on
ld a, [rLCDC]
or a, LCDCF_OBJON
ldh [rLCDC], a
```

> **To turn the Sprites Off:**

Reviewer 4: ```suggestion
**To turn sprite objects off:**
```

```rgbasm,linenos
; Turn the sprites off
ld a, [rLCDC]
and a, LCDCF_OBJOFF
ldh [rLCDC], a
```

::: tip

> Sprites by default are in 8x8 mode.

Reviewer 4: ```suggestion
Sprites are in 8x8 mode by default.
```

:::

### Enable tall (8x16) sprites

> Once sprites are enabled, you can enable tall sprites using the third least signficiant bit of the `rLCDC` register: `LCDCF_OBJ16`

Reviewer 4: ```suggestion
Once sprites are enabled, you can enable tall sprites using the 3rd bit of the `rLCDC` register: `LCDCF_OBJ16`
```

::: tip

You can not have some 8x8 sprites and some 8x16 sprites. All sprites must be of the same size.

:::

```rgbasm,linenos
; Turn tall sprites on
ld a, [rLCDC]
or a, LCDCF_OBJ16
ldh [rLCDC], a
```

## Backgrounds

### Put background/window tile data into VRAM

> The reigon in VRAM dedicated for the background/window tilemaps is from $9000 to $97F0. Harware.inc defines a `_VRAM9000` constant you can use for that. To copy background/window tile data into VRAM, you can use a loop to copy the bytes.

Reviewer 5: spelling
```suggestion
The region in VRAM dedicated for the background/window tilemaps is from $9000 to $97F0. Hardware.inc defines a `_VRAM9000` constant you can use for that. To copy background or window tile data into VRAM, you can use a loop to copy the bytes.
```

> The region in VRAM dedicated for the background/window tilemaps is from $9000 to $97F0. Hardware.inc defines a `_VRAM9000` constant you can use for that. To copy background or window tile data into VRAM, you can use a loop to copy the bytes.

Reviewer 4: I don't think mentioning *how* to copy bytes is necessary here.
```suggestion
The region in VRAM dedicated for the background/window tilemaps is from $9000 to $97FF. hardware.inc defines a `_VRAM9000` constant you can use for that.
```

> myBackground: INCBIN "src/path/to/my/my-background.2bpp"
> myBackgroundEnd:
> 
> CopyBackgroundWindowTileDataIntoVram:
> ; Copy the tile data
> ld de, myBackground
> ld hl, \_VRAM
> ld bc, myBackgroundEnd - myBackground
> CopyBackgroundWindowTileDataIntoVram_Loop:
> ld a, [de]
> ld [hli], a
> inc de
> dec bc
> ld a, b
> or a, c
> jp nz, CopyBackgroundWindowTileDataIntoVram_Loop

Reviewer 4: ```suggestion
MyBackground:
    INCBIN "src/path/to/my-background.2bpp"
.end

CopyBackgroundWindowTileDataIntoVram:
    ; Copy the tile data
    ld de, myBackground
    ld hl, \_VRAM
    ld bc, MyBackground.end - MyBackground
.loop:
    ld a, [de]
    ld [hli], a
    inc de
    dec bc
    ld a, b
    or a, c
    jr nz, .Loop
```

### Draw on the Background/Window

The Game Boy has 2 32x32 tilemaps, one at `$9800` and another at `$9C00`. Either can be used for the background or window. By default, they both use the tilemap at `$9800`.

Drawing on the background or window is as simple as copying bytes starting at one of those addresses:

```rgbasm, lineno
CopyTilemapTo
   ; Copy the tilemap
    ld de, Tilemap
    ld hl, $9800
    ld bc, TilemapEnd - Tilemap
CopyTilemap:
    ld a, [de]
    ld [hli], a
    inc de
    dec bc
    ld a, b
    or a, c
    jp nz, CopyTilemap
```

::: tip

> Make sure the layer you're targetting has been turned on. See ["How to turn on/off the window"](#how-to-turn-onoff-the-window) and ["How to turn on/off the background"](#how-to-turn-onoff-the-background)

Reviewer 3: broken references ("how to ...")

:::

::: tip

In terms of tiles, The background/window tilemaps are 32x32. The Game Boy's screen is 20x18. When copying tiles, understand that RGBDS or the Game Boy won't automatically jump to the next visible row after you've reached the 20th column.

:::

### Move the background

You can move the background horizontally & vertically using the `$FF43` and `$FF42` registers, respectively. Hardware.inc defines two constants for that: `rSCX` and `rSCY`.

**How to change the background's X Position:**

```rgbasm,linenos
ld a,64
ld [rSCX], a
```

**How to change the background's Y Position:**

```rgbasm,linenos
ld a,64
ld [rSCY], a
```

Check out the Pan Docs for more info on the [Background viewport Y position, X position](https://gbdev.io/pandocs/Scrolling.html#ff42ff43--scy-scx-background-viewport-y-position-x-position)

### Move the window

Moving the window is the same as moving the background, except using the `$FF4B` and `$FF4A` registers. Hardware.inc defines two constants for that: `rWX` and `rWY`.

::: tip

> The window layer has an -8 pixel horizontal offset.

Reviewer 5: object offset is 8, and window offset is 7
```suggestion
The window layer has a -7 pixel horizontal offset. This means setting `rWX` to 7 places the window at the left side of the screen, and setting `rWX` to 87 places the window with its left side halfway across the screen.
```

```

### Wait for a button press

To wait **indefinitely** for a button press, create a loop where you:

-   check if the button has JUST been pressed
-   If not:
    -   Wait until the next vertical blank phase completes
    -   call the `UpdateKeys` function again
    -   Loop background to the beginning

::: tip

This will halt all other logic (outside of interrupts), be careful if you need any logic running simultaneously.

:::

```rgbasm, linenos
WaitForAButtonToBePressed:
    ld a, [wNewKeys]
    and a, PADF_A
    ret nz
WaitUntilVerticalBlankStart:
    ld a, [rLY]
    cp 144
    jp nc, WaitUntilVerticalBlankStart
WaitUntilVerticalBlankEnd:
    ld a, [rLY]
    cp 144
    jp c, WaitUntilVerticalBlankEnd
    call UpdateKeys
    jp WaitForAButtonToBePressed
```

## HUD

Heads Up Displays, or HUDs; are commonly used to prevent extra information to the player. Good examples are: Score, Health, and the current level. The window layer is drawn on top of the background, and cannot move like the background. For this reason, commonly the window layer is used for HUDs. See ["How to Draw on the Background/Window"](#how-to-draw-on-the-backgroundwindow).

### Draw text

> Drawing text on the window is essentially drawing tiles (with letters/numbers/puncatuation on them) on the window and/or background layer.

Reviewer 5: spelling
```suggestion
Drawing text on the window is essentially drawing tiles (with letters/numbers/punctuation on them) on the window and/or background layer.
```

To simplify the process you can define constant strings.

::: tip

> These constants end with a literall 255, which our code will read as the end of the string.

Reviewer 5: spelling
```suggestion
These constants end with a literal 255, which our code will read as the end of the string.
```

:::

```rgbasm, lineno

SECTION "Text ASM", ROM0

wScoreText::  db "score", 255

```

RGBDS has a character map functionality. You can read more in the [RGBDS Assembly Syntax Documentation](https://rgbds.gbdev.io/docs/v0.6.1/rgbasm.5#DEFINING_DATA). This functionality, tells the compiler how to map each letter:

::: tip

You need to have your text font tiles in VRAM at the locations specified in the map. See [How to put background/window tile data in VRAM](#how-to-put-backgroundwindow-tile-data-into-vram)

:::

```rgbasm, lineno

CHARMAP " ", 0
CHARMAP ".", 24
CHARMAP "-", 25
CHARMAP "a", 26
CHARMAP "b", 27
CHARMAP "c", 28
CHARMAP "d", 29
CHARMAP "e", 30
CHARMAP "f", 31
CHARMAP "g", 32
CHARMAP "h", 33
CHARMAP "i", 34
CHARMAP "j", 35
CHARMAP "k", 36
CHARMAP "l", 37
CHARMAP "m", 38
CHARMAP "n", 39
CHARMAP "o", 40
CHARMAP "p", 41
CHARMAP "q", 42
CHARMAP "r", 43
CHARMAP "s", 44
CHARMAP "t", 45
CHARMAP "u", 46
CHARMAP "v", 47
CHARMAP "w", 48
CHARMAP "x", 49
CHARMAP "y", 50
CHARMAP "z", 51

```

The above character mapping would convert (by the compiler) our `wScoreText` text to:

-   s => 44
-   c => 28
-   o => 40
-   r => 43
-   e => 30
-   255

> With that setup, we would loop though the bytes of `wScoreText`. While drawing copying them over to the background/window layer. After we copy each byte, we'll increment where we will copy to, and which byte in `wScoreText` we are reading. When we read 255, our code will end.

Reviewer 5: sentence fragment
```suggestion
With that setup, we would loop though the bytes of `wScoreText` and copy each byte to the background/window layer. After we copy each byte, we'll increment where we will copy to, and which byte in `wScoreText` we are reading. When we read 255, our code will end.
```

::: tip

This example implies that your font tiles are located in VRAM at the locations specified in the character mapping.

:::

** Drawing 'score' on the window **

```rgbasm, lineno

DrawTextTiles::

    ld hl, wScoreText
    ld de, $9C00 ; The window tilemap starts at $9C00

DrawTextTilesLoop::

    ; Check for the end of string character 255
    ld a, [hl]
    cp 255
    ret z

    ; Write the current character (in hl) to the address
    ; on the tilemap (in de)
    ld a, [hl]
    ld [de], a

    inc hl
    inc de

    ; move to the next character and next background tile
    jp DrawTextTilesLoop
```

### Draw a bottom HUD

-   Enable the window (with a different tilemap than the background)
-   Move the window downwards, so only 1 or 2 rows show at the bottom of the screen
-   Draw your text, score, and icons on the top of the window layer.

::: tip

Sprites will still show over the window. To fully prevent that, you can use STAT interrupts to hide sprites where the bottom HUD will be shown.

:::

## Sprites

### Put sprite tile data in VRAM

> The reigon in VRAM dedicated for sprites is from `$8000` to `$87F0`. Harware.inc defines a `_VRAM` constant you can use for that. To copy sprite tile data into VRAM, you can use a loop to copy the bytes.

Reviewer 5: spelling
```suggestion
The region in VRAM dedicated for sprites is from `$8000` to `$87F0`. Hardware.inc defines a `_VRAM` constant you can use for that. To copy sprite tile data into VRAM, you can use a loop to copy the bytes.
```

CopySpriteTileDataIntoVram_Loop:
    ld a, [de]
    ld [hli], a
    inc de
    dec bc
    ld a, b
    or a, c
    jp nz, CopySpriteTileDataIntoVram_Loop
```

### Manipulate hardware OAM sprites

Each hardware sprite has 4 bytes: (in this order)

-   Y position
-   X Position
-   Tile ID
-   Flags/Props (priority, y flip, x flip, palette 0 [DMG], palette 1 [DMG], bank 0 [GBC], bank 1 [GBC])

Check out the Pan Docs page on [Object Attribute Memory (OAM)](https://gbdev.io/pandocs/OAM.html) for more info.

The bytes controlling hardware OAM sprites start at `$FE00`, for which hardware.inc has defined a constant as `_OAMRAM`.

**Moving (the first) OAM sprite, one pixel downwards:**

```rgbasm, linenos
ld a, [_OAMRAM]
inc a
ld [_OAMRAM], a
```

**Moving (the first) OAM sprite, one pixel to the right:**

```rgbasm, linenos
ld a, [_OAMRAM + 1]
inc a
ld [_OAMRAM + 1], a
```

**Setting the tile for the first OAM sprite:**

```rgbasm, linenos
ld a, 3
ld [_OAMRAM+2], a
```

**Moving (the fifth) OAM sprite, one pixel downwards:**

```rgbasm, linenos
ld a, [_OAMRAM + 20]
inc a
ld [_OAMRAM + 20], a
```

TODO - Explanation on limitations of direct OAM manipulation.

::: tip

> It's recommended developers implement a shadow OAM, like @eievue5's [Sprite Object Library](https://github.com/eievui5/gb-sprobj-lib)

Reviewer 5: spelling
```suggestion
It's recommended that developers implement a shadow OAM, like @eievui5's [Sprite Object Library](https://github.com/eievui5/gb-sprobj-lib)
```

:::

> ### Iimplement a Shadow OAM using @eievue5's Sprite Object Library
> 
> Github URL: [https://github.com/eievui5/gb-sprobj-lib](https://github.com/eievui5/gb-sprobj-lib)

Reviewer 5: spelling
```suggestion
### Implement a Shadow OAM using @eievui5's Sprite Object Library

GitHub URL: [https://github.com/eievui5/gb-sprobj-lib](https://github.com/eievui5/gb-sprobj-lib)
```

> This is a small, lightweight library meant to facilitate the rendering of sprite objects, including Shadow OAM and OAM DMA, single-entry "simple" sprite objects, and Q12.4 fixed-point position metasprite rendering.

**Usage**

The library is relatively simple to get set up. First, put the following in your initialization code:

```rgbasm, linenos
    ; Initilize Sprite Object Library.
    call InitSprObjLib

    ; Reset hardware OAM
    xor a, a
    ld b, 160
    ld hl, _OAMRAM
.resetOAM
    ld [hli], a
    dec b
    jr nz, .resetOAM
```

Then put a call to `ResetShadowOAM` at the beginning of your main loop.

Finally, run the following code during VBlank:

```rgbasm, linenos
ld a, HIGH(wShadowOAM)
call hOAMDMA
```

### Manipulate Shadow OAM OAM sprites

> Once you've setup @eievue5's Sprite Object Library, you can manipulate shadow OAM sprites the exact same way you would manipulate normal hardware OAM sprites. Except, this time you would use the librarys `wShadowOAM` constant instead of the `_OAMRAM` register.

Reviewer 5: spelling
```suggestion
Once you've set up @eievui5's Sprite Object Library, you can manipulate shadow OAM sprites the exact same way you would manipulate normal hardware OAM sprites. Except, this time you would use the library's `wShadowOAM` constant instead of the `_OAMRAM` register.
```

**Moving (the first) OAM sprite, one pixel downwards:**

```rgbasm, linenos
ld a,LOW(wShadowOAM)
ld l, a
ld a, HIGH(wShadowOAM)
ld h, a

ld a, [hl]
inc a
ld [wShadowOAM], a
```

## Micelaneous

### Save Data

> If you want to save data in your game your game's header needs to specify the correct mbc/cartridge type, and it needs to have a non-zero SRAM size. This should be done in your makefile by passing special parameters to [rgbfix](https://rgbds.gbdev.io/docs/v0.6.1/rgbfix.1).

Reviewer 5: capitalize initialism
```suggestion
If you want to save data in your game, your game's header needs to specify the correct MBC/cartridge type, and it needs to have a non-zero SRAM size. This should be done in your makefile by passing special parameters to [rgbfix](https://rgbds.gbdev.io/docs/v0.6.1/rgbfix.1).
```

-   Use the `-m` or `--mbc-type` parameters to set the mbc/cartidge type, 0x147, to a given value from 0 to 0xFF. [More Info](https://gbdev.io/pandocs/The_Cartridge_Header.html#0147--cartridge-type)
-   Use the `-r` or `--ram-size` parameters to set the RAM size, 0x149, to a given value from 0 to 0xFF. [More Info](https://gbdev.io/pandocs/The_Cartridge_Header.html#0149--ram-size).

To save data you need to store variables in Static RAM. This is done by creating a new SRAM "SECTION". [More Info](https://rgbds.gbdev.io/docs/v0.6.1/rgbasm.5#SECTIONS)

```rgbasm, linenos
SECTION "SaveVariables", SRAM

wCurrentLevel:: db

```

To access SRAM, you need to write `CART_SRAM_ENABLE` to the `rRAMG` register. When done, you can disable SRAM using the `CART_SRAM_DISABLE` constant.

**To enable read/write access to SRAM:**

```rgbasm, linenos

ld a, CART_SRAM_ENABLE
ld [rRAMG], a

```

**To disable read/write access to SRAM:**

```rgbasm, linenos

ld a, CART_SRAM_DISABLE
ld [rRAMG], a

```

**Initiating Save Data**

By default, save data for your game may or may not exist. When the save data does not exist, the value of the bytes dedicated for saving will be random.

> You can dedicate a couple bytes towards creating a pseduo-checksum. When these bytes have a **very specific** value, you can be somewhat sure the save data has been initialized.

Reviewer 5: spelling
```suggestion
You can dedicate a couple bytes towards creating a pseudo-checksum. When these bytes have a **very specific** value, you can be somewhat sure the save data has been initialized.
```

```rgbasm, linenos
SECTION "SaveVariables", SRAM

wCurrentLevel:: db
wCheckSum1:: db
wCheckSum2:: db
wCheckSum3:: db
```

> When initiating your save data, you'll need to

Reviewer 5: initiate means to perform a ceremony to add a member to a club
```suggestion
When initializing your save data, you'll need to
```

-   enable SRAM access
-   set your checksum bytes
-   give your other variables default values
-   disable SRAM access

```rgbasm, linenos

;; Setup our save data
InitSaveData::

    ld a, CART_SRAM_ENABLE
    ld [rRAMG], a

    ld a, 123
    ld [wCheckSum1], a

    ld a, 111
    ld [wCheckSum2], a

    ld a, 222
    ld [wCheckSum3], a

    ld a, 0
    ld [wCurrentLevel], a

    ld a, CART_SRAM_DISABLE
    ld [rRAMG], a

    ret
```

> Once your save file has been iniated, you can access any variable normally once SRAM is enabled.

Reviewer 5: spelling
```suggestion
Once your save file has been initialized, you can access any variable normally once SRAM is enabled.
```

```rgbasm, linenos

;; Setup our save data
StartNextLevel::

    ld a, CART_SRAM_ENABLE
    ld [rRAMG], a

    ld a, [wCurrentLevel]
    cp a, 3
    call z, StartLevel3

    ld a, CART_SRAM_DISABLE
    ld [rRAMG], a

    ret
```

### Generate random numbers

> Random number generation is a [complex task in software](https://en.wikipedia.org/wiki/Random_number_generation). What you can implement is a "pseudorandom" generator, giving you a very unpredictable sequence of values. Here's a `rand` function (from [Damien Yerrick](https://github.com/pinobatch)) you can use.

Reviewer 5: spelling
```suggestion
Random number generation is a [complex task in software](https://en.wikipedia.org/wiki/Random_number_generation). What you can implement is a "pseudorandom" generator, giving you a very unpredictable sequence of values. Here's a `rand` function (from [Damian Yerrick](https://github.com/pinobatch)) you can use.
```
